### ISSUE TITLE
Enhancement for media streaming

### ISSUE THREAD

**Issue by yongtang** [tensorflow/io#505]
From https://github.com/tensorflow/io/issues/453#issuecomment-534605676

@yongtang Thank you for keeping me in the loop. As I mentioned earlier, I've spent more time looking at the options for real time media streaming feeds into TF IO. Seems like we are independently reaching similar conclusions. ffmpeg works well for transforming to raw media a large number of file containers and formats, but it lacks fine grained control, which is especially important for capturing and replaying real time media with all of its network characteristics like dropped packets, latency, jitter, etc. Its probably best to split the discussion  into several buckets:
1. Basic use case of reading from an arbitrary recorded media file into TFIO. This does not deal with network artifacts. Applicable for scenarios where the media is pre-recorded. Since the file is static, the same original source can be reused many times for training.
2. Real time media streaming for inference. The media is live and has network artifacts, but does not need to be recorded for training purpose. The aim here is real time inference performance.
3. Real time media streaming for training. The media has network artifacts that need to be precisely captured (e.g. via pcap) and organized for model training. The emphasis here is on precision in capturing all important dataset features. Not just the packets carrying media signal, but also the network delays and other artifacts that can impact the Bayesian baseline banchmark of a model.

I think for 1. the recent PRs have made a ton of good progress and the topic can be soon closed. For 2&3, I am researching ways forward with tools like gstreamer, but don't have results to report yet.

**Comment by yongtang** [comment]
@ivelin Yes for audio and video format support, FFmpeg provides almost everything. However, for machine learning purposes I think user only care about several limited formats that are the most widely used. Less widely used formats could always be converted (since they are rare).

One note is that, patent/license could be an issue for some codec. So other than one widely used format (like H264), ideally tensorflow-io should support an open source alternative as well. Overall,  formats should be covered by tensorflow-io for audio/video are truly limited:

 - MP4 for Audio/Video and associated codec - That is the most widely seen format. It may also have patent/license issues. On Windows/macOS I believe patent/license are properly covered by native system APIs (e.g., Apple's AVFoundation). On Linux FFmpeg could be used but patent are not covered.
- WAV - Widely used Audio formats without patent limitation. We already have limited support in tensorflow-io.
- WebM - Open source alternative
- Ogg - Open source alternative

For tensorflow-io, I think additional work could be limited to using native system APIs on macOS/Windows for patent covered codec, and adding open source formats natively (with indexable/random access to replace FFmpeg).

**Comment by yongtang** [comment]
Also, subtitle could play an important role in audio/training, as subtitle itself is a label with associated feature in audio.

With subtitle it could be combined for training audio/video data in an easy way.

PR #499 adds the Subtitle through FFmpeg. Though Ffmpeg's support for WebVTT (an html5 standard that are normally embedded in web pages together with html5 video) seems to be subpar.

**Comment by yongtang** [comment]
@ivelin For 2), wondering if bypass an intermediate gstreamer makes sense? Basically, once raw audio data is captured from device input, it will be serialized and send through grpc directly. (or have a preprocessing that could happen before the data is sent).

**Comment by ivelin** [comment]
Yes, raw input from the capture device is certainly an option whenever its
possible to run a GRPC client for TFIO on the device.
